test/framework/control_plane.go

```diff
@@ -34,213 +34,276 @@ import (
-		Eventually(func() error {
-			return mgmtClient.Create(ctx, obj)
-		}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
-	}
+// CreateKubeadmControlPlaneInput is the input for CreateKubeadmControlPlane.
+type CreateKubeadmControlPlaneInput struct {
+	Creator         Creator
+	ControlPlane    *controlplanev1.KubeadmControlPlane
+	MachineTemplate runtime.Object
+}
 
+// CreateKubeadmControlPlane creates the control plane object and necessary dependencies.
+func CreateKubeadmControlPlane(ctx context.Context, input CreateKubeadmControlPlaneInput, intervals ...interface{}) {
 	By("creating the machine template")
-	Expect(mgmtClient.Create(ctx, input.MachineTemplate)).To(Succeed())
+	Expect(input.Creator.Create(ctx, input.MachineTemplate)).To(Succeed())
 
 	By("creating a KubeadmControlPlane")
 	Eventually(func() error {
-		err := mgmtClient.Create(ctx, input.ControlPlane)
+		err := input.Creator.Create(ctx, input.ControlPlane)
 		if err != nil {
 			fmt.Println(err)
 		}
 		return err
-	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+	}, intervals...).Should(Succeed())
+}
+
+// CreateMachineDeploymentInput is the input for CreateMachineDeployment.
+type CreateMachineDeploymentInput struct {
+	Creator                 Creator
+	MachineDeployment       *clusterv1.MachineDeployment
+	BootstrapConfigTemplate runtime.Object
+	InfraMachineTemplate    runtime.Object
+}
+
+// CreateMachineDeployment creates the machine deployment and dependencies.
+func CreateMachineDeployment(ctx context.Context, input CreateMachineDeploymentInput) {
+	By("creating a core MachineDeployment resource")
+	Expect(input.Creator.Create(ctx, input.MachineDeployment)).To(Succeed())
+
+	By("creating a BootstrapConfigTemplate resource")
+	Expect(input.Creator.Create(ctx, input.BootstrapConfigTemplate)).To(Succeed())
+
+	By("creating an InfrastructureMachineTemplate resource")
+	Expect(input.Creator.Create(ctx, input.InfraMachineTemplate)).To(Succeed())
+}
+
+// WaitForMachineDeploymentNodesToExistInput is the input for WaitForMachineDeploymentNodesToExist.
+type WaitForMachineDeploymentNodesToExistInput struct {
+	Management        ManagementCluster
+	Cluster           *clusterv1.Cluster
+	MachineDeployment *clusterv1.MachineDeployment
+}
 
+// WaitForMachineDeploymentNodesToExist waits until all nodes associated with a machine deployment exist.
+func WaitForMachineDeploymentNodesToExist(ctx context.Context, input WaitForMachineDeploymentNodesToExistInput, intervals ...interface{}) {
+	By("waiting for the workload nodes to exist")
+	Eventually(func() ([]v1.Node, error) {
+		workloadClient, err := input.Management.GetWorkloadClient(ctx, input.Cluster.Namespace, input.Cluster.Name)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get workload client")
+		}
+		nodeList := v1.NodeList{}
+		if err := workloadClient.List(ctx, &nodeList); err != nil {
+			return nil, err
+		}
+		return nodeList.Items, nil
```

Comment:
Your suggestion would definitely be an improvement. It would be more true to cluster-api's vision of the cluster.

However this is what the test was doing previous. I'll open an issue to update to improve this function.

https://github.com/kubernetes-sigs/cluster-api/issues/2309

Reply:
Agree with deferring the improvement to checking Machines, however Control Plane machines will also have associated Nodes on the remote cluster, and this currently is trying to compare the number of nodes retrieved against the number of desired replicas for the MachineDeployment.

Reply:
ah, ok, this requires a change then. The assumption that this PR fixes unfortunately exposed this bug. Good code reading!

Reply:
:) maybe i meant fortunately.